### DISCUSSION TITLE
Conditional Retry

### DISCUSSION THREAD

**Discussion by julianwachholz** [TanStack/query#84] [General] [answered]
I'd like to explore the possibility of adding a conditional retry feature.

One might not want to retry if the server responds with a 404, then we just want to display a "not found" page. But we still want to retry if there's a temporary error somewhere else.

**Comment by nikoskleidis** [comment]
If I understand correctly what you are trying to do, you can use `retry` option as described here:
https://github.com/tannerlinsley/react-query#retries

**Comment by julianwachholz** [comment]
Thanks for your input, I am using this setting but it doesn't allow conditional values as I've explained.

**Comment by nikoskleidis** [comment]
Then I think you should handle the 404 error internally (without throwing an error). You can throw an error only for status codes that you want react-query to retry fetching.

**Comment by julianwachholz** [comment]
You're right, I'm silly. Right now I'm just throwing when `!response.ok`, which might be a bit overzealous.

**Comment by tannerlinsley** [comment]
Let me know if there is any more to discuss here.

**Comment by smozely** [comment]
Don't know if I should raise a new issue or continue the discussion here ... 

The idea for handling the errors internally doesn't work that well for using Suspense and Error boundaries.

What we thinking we want to do is have failed requests be sent up the error boundaries, and there we would have a error boundary that handles just the data being unavailable, and a boundary higher up the stack that handles more fatal things ... like session timeouts, which in our case need to send the user to our login application. And here a session timeout error would be something we don't want to retry on.

In a POC of writing our own lib for handling this type of stuff, we had a magic "NoRetryError" which would wrap a real error, that we would throw to indicate something shouldn't be retried, which worked ... but felt a little gross because it was the only time our fetching stuff needed to know it was being wrapped in our lib. Another thing we considered was adding a magic prop to an Error to indicate not to retry ... but has that same issue of needing to know its being wrapped.

**Comment by tannerlinsley** [comment]
So what exactly are you saying needs to happen?

**Reply by smozely** [reply]
So I think I need to be able to throw an error which tells react-query not to retry

Either ...
* A config function that can inspect the error and return true / false if it should be retried
* A `NoRetryError` type the the fetch can throw
* Adding something like `.noRetry = true` before the fetch throws

I'm thinking the config function might be best

**Reply by smozely** [reply]
Actually I've just looked at the code, 

I think it would be easy to do the version of a config function. theres just that spot where it checks if it should retry, and there we can call the function, which can decide if its a retry-able error or not ... the default being true.

I'd be happy to submit a PR for that, as long as you don't think its a bad idea and would never merge it.

**Reply by julianwachholz** [reply]
The function to check if it should retry was my original thought to this as well.

**Comment by cherniavskii** [comment, marked as answer]
Implemented in https://github.com/tannerlinsley/react-query/pull/294

**Reply by steveoh** [reply]
I really like the way that ky did this with retry status codes. I think this would be an improvement for react-query.

statusCodes: 408 413 429 500 502 503 504

**Reply by TkDodo** [reply]
please keep in mind that react query is promise-based and _not_ necessarily a network library. It doesn't know about status codes, because data can come from other async sources as well.

**Reply by satyajitnayk** [reply]
I think this could be done @julianwachholz. Conditional Retry in react-query
Try this on [playcode](https://playcode.io/react)

**App.jsx**
```js
import React from 'react';
import { useQuery, QueryClient, QueryClientProvider } from 'react-query';

// Mock API fetch function
const fetchData = () => {
  return new Promise((resolve, reject) => {
    setTimeout(() => {
      const randomError = Math.random() > 0.3;
      if (randomError) {
        const serverError = Math.random() > 0.5;
        reject({ status: serverError ? 500 : 400 ,message:  'Error occurred'});
      } else {
        resolve({ data: 'Success!' });
      }
    }, 1000);
  });
};

// React component using useQuery
const MyComponent = () => {
  const { data, error, isLoading } = useQuery('data', fetchData, {
    retry: (failureCount, error) => {
       console.log(`Response Code: ${error.status} failureCount: ${failureCount}`);
       const shouldRetry = error.status === 500 && failureCount < 3;
       if(shouldRetry) console.log('Retrying ....!');
      // retry on 500 errors for max 3 times
      return shouldRetry;  
    },
  });

  if (isLoading) return <div>Loading...</div>;
  if (error) {
    return <div>{error.message}</div>;
  }

  return <div>{JSON.stringify(data)}</div>;
};

// Setup QueryClient and wrap your component with QueryClientProvider
const queryClient = new QueryClient();

const App = () => (
  <QueryClientProvider client={queryClient}>
    <MyComponent />
  </QueryClientProvider>
);

export default App;
```

<img width="1298" alt="image" src="https://github.com/TanStack/query/assets/32722867/69de2e6e-18e5-45ef-a3a4-838f636763d6">

**Reply by ciprianmacovei** [reply]
@TkDodo yes it s not but i pass to this lib the request, u can take the status code right from that ?, why not passing it to the retry fuction like (retryCounts, status) => 

If i pass to u the request and u consume it, probably u can get the status code as well

**Reply by TkDodo** [reply]
> why not passing it to the retry fuction like (retryCounts, status) =>

react query is not tied to data fetching, so it doesn’t know about http status

**Reply by steveoh** [reply]
It could be abstracted with a convention of status types similar to the HTTP spec that should satisfy most if not all use cases?

**Reply by TkDodo** [reply]
Sorry, not gonna happen. This needs to work:

```
useQuery({
  queryKey: ['answer'],
  queryFn: () => Promise.resolve(42)
}
```

there’s no http call, not status, no object returned. Just a number. If you want to return `{ data, headers, status }` from your `queryFn`, that’s something _your_ API layer / abstraction needs to do.

Then, you can implement retry by looking at `data.status`

**Reply by ciprianmacovei** [reply]
yes but the promise can return the response obj right ? u can enforce ppl to return the response obj not the data or whateva, because this library is made for reqests , a wrapper for req right ?

**Reply by ciprianmacovei** [reply]
and data will be the response obj, and can be destructed, i'm only proposing 😅

**Reply by TkDodo** [reply]
> this library is made for reqests , a wrapper for req right ?

it’s for async state - we only enforce that a Promise must be returned. Doesn’t have to be an object with `data` property or whatever. If you think we can make a (breaking) change like that you’re grossly underestimating the impact that would have on millions of existing usages for little gain.

**Reply by ciprianmacovei** [reply]
yeah ur right
